### PR TITLE
Add ability to login via email

### PR DESCRIPTION
**Pull request by Gauravjeetsingh**
- Create new game sessions
- Add ability to login via email

### CODE REVIEW

File: src/app/api/session/route.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+import { Progress, SessionData } from "@/schema";
+import { defaultDB, usersDb } from "@/lib/firebaseAdmin";
+import { shuffleElements } from "@/components/game-session/utils/shuffle";
+
+import { addQuestionsToSession, addSlidesToSession } from "./addToSession";
+import { getWordDefinitions } from "./getWordDefinitions";
+import { createQuestionFromList } from "./createQuestionFromList";
+
+export async function POST(req: Request) {
```
Comment (gurjit03) [inline src/app/api/session/route.ts:11]: @Gauravjeetsingh This is almost 200 lines function. Can we divide this into multiple pieces. This 1 route seems doing a lot of things is not so readable 🙏🏻 .

**Review by gurjit03 — COMMENTED** [review]
I am able to review half of it.

**Comment by amitojsingh** [comment]
I am not able to see session it's stuck on verification. I verified and refresh the page nothing happens. 
Email I am using is cideh72017@ploncy.com

File: src/app/api/session/addToSession.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+              options: slide.options,
+              word: slide.word,
+              image: slide.image,
+            },
+          };
+        default:
+          return [];
+      }
+    });
+    const questionIndexs = sessionArray.reduce((indices: number[], obj, index: number) => {
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/addToSession.ts:56]: fixed

File: src/app/api/session/addToSession.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        case "definition":
+          return {
+            key: `definition-${slide.id}`,
+            questionCount: 0,
+            data: {
+              word: slide.word,
+              translation: slide.translation,
+              meaningEnglish: slide.meaning_english,
+              meaningPunjabi: slide.meaning_punjabi,
+              images: slide.images,
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/addToSession.ts:23]: yeah, there are multiple images in some word definitions

File: src/app/api/session/createQuestionFromWord.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { Progress } from "@/schema";
+import { fetchQuestions } from "./fetchQuestions";
+
+export const createQuestionFromWord = async (
+  learnedWord: Progress,
+  questionCollection: FirebaseFirestore.CollectionReference
+) => {
+  let questionData: Array<FirebaseFirestore.DocumentData> = [];
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/createQuestionFromWord.ts:8]: Changed it to [] everywhere.

File: src/app/api/session/createQuestionFromList.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { Progress } from "@/schema";
+import { createQuestionFromWord } from "./createQuestionFromWord";
+
+export const createQuestionFromList = async (
+  learningWords: Progress[],
+  questionCollection: FirebaseFirestore.CollectionReference
+) => {
+  const learningWordQuestions: FirebaseFirestore.DocumentData[] = [];
+  for (const wordObj of learningWords) {
+   const data = await createQuestionFromWord(wordObj, questionCollection);
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/createQuestionFromList.ts:10]: fixed

File: src/app/api/session/createQuestionFromWord.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+  const { questionIds, wordId, wordText } = learnedWord;
+  const chunkSize = 10;
+
+  if (questionIds.length >= chunkSize) {
+    const sublist = questionIds.slice(0, chunkSize);
+    const initialData = await fetchQuestions(sublist, questionCollection, wordId, wordText);
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/createQuestionFromWord.ts:15]: Updated it to allQuestions. 👍

File: src/app/api/session/fetchQuestions.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+export const fetchQuestions = async (
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/fetchQuestions.ts:1]: yeah, agreed. Updated.

File: src/app/api/session/getWordDefinitions.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  const excludedWordIds = wordIds;
+  if (excludedWordIds.length >= 10) {
+    const sublist = excludedWordIds.splice(0, 10);
+    const wordQuery = await wordsCollection
+      .where("status", "==", "active")
+      .where("id", "not-in", sublist)
+      .get();
+
+    const wordData = wordQuery.docs.map((doc) => doc.data());
+    const filteredData = wordData.filter(
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply src/app/api/session/getWordDefinitions.ts:15]: Renamed filteredData to filteredWordData

File: src/app/api/session/route.ts (hunk 1 above)
Comment (Gauravjeetsingh) [reply src/app/api/session/route.ts:11]: Yeah, this is the main file that is creating the game session, we can look into this later, maybe as a separate PR for this refactor.

**Pull request merged**